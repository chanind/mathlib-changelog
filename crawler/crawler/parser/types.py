from __future__ import annotations
from dataclasses import dataclass
from typing import Literal


ItemType = Literal["theorem", "def", "inductive", "structure"]
BlockType = Literal["section", "namespace"]


@dataclass
class ParsedItem:
    type: ItemType
    name: str
    namespace: list[str]
    line_hash: str

    @property
    def full_name(self) -> str:
        return self.type + ":" + ".".join([*self.namespace, self.name])


@dataclass
class LeanParseResult:
    items: list[ParsedItem]

    def get_item(self, full_name: str) -> ParsedItem | None:
        for item in self.items:
            if item.full_name == full_name:
                return item
        return None


@dataclass
class ParseState:
    blocks: list[tuple[BlockType, str]]
    items: list[ParsedItem]

    @property
    def namespace(self) -> list[str]:
        namespace = []
        for block in self.blocks:
            if block[0] == "namespace":
                # handle cases like "namespace Thing1.thing2"
                namespace += block[1].split(".")
        return namespace
