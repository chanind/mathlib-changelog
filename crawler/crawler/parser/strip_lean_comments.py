import re


def strip_lean_comments(lean_contents: str) -> str:
    processed = re.sub(r"/-((?!-/)(.|\r|\n))*-/", "", lean_contents, re.MULTILINE)
    return re.sub(r"\-\-[^\n\r]*", "", processed, re.MULTILINE)
