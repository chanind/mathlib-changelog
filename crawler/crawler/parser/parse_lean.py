from __future__ import annotations
from dataclasses import dataclass
import logging
import re
from typing import Literal, cast
import hashlib

from .errors import LeanParseError
from .strip_lean_comments import strip_lean_comments


ItemType = Literal["lemma", "theorem", "def"]
BlockType = Literal["section", "namespace"]


@dataclass
class ParsedItem:
    type: ItemType
    name: str
    namespace: list[str]
    line_hash: str

    @property
    def full_name(self) -> str:
        return self.type + ":" + ".".join([*self.namespace, self.name])


@dataclass
class LeanParseResult:
    items: list[ParsedItem]

    def get_item(self, full_name: str) -> ParsedItem | None:
        for item in self.items:
            if item.full_name == full_name:
                return item
        return None


@dataclass
class ParseState:
    blocks: list[tuple[BlockType, str]]
    items: list[ParsedItem]

    @property
    def namespace(self) -> list[str]:
        namespace = []
        for block in self.blocks:
            if block[0] == "namespace":
                # handle cases like "namespace Thing1.thing2"
                namespace += block[1].split(".")
        return namespace


def parse_lean(contents: str) -> list[ParsedItem]:
    state = ParseState(blocks=[], items=[])

    lines = strip_lean_comments(contents).splitlines()
    for line in lines:
        parse_line(line, state)
    if len(state.namespace) > 0:
        logging.warning(
            f"Invalid parse state! open namespace remaining at end of parse: {state.namespace}"
        )
    return state.items


OPTIONAL_DECORATOR_REGEX = r"(?:@\[[^\]]+\]\s?)"
ITEM_REGEX = rf"^\s*{OPTIONAL_DECORATOR_REGEX}*(def|lemma|theorem)\s+([^\s]+)"
START_BLOCK_REGEX = r"^\s*(section|namespace)\s+([^\s]+)"
END_BLOCK_REGEX = r"^\s*end\s+([^\s]+)"


def parse_line(line: str, state: ParseState) -> None:
    """
    Extract any items from the line, and track sections/namespaces.
    This will modify the state in-place.

    NOTE: this assumes there can only be a single def/theorem/lemma/section/namespace per line
    """
    item_match = re.search(ITEM_REGEX, line)
    start_block_match = re.search(START_BLOCK_REGEX, line)
    end_block_match = re.search(END_BLOCK_REGEX, line)

    if item_match:
        line_hash = hashlib.md5(line.encode()).hexdigest()
        [item_type, item_name] = item_match.groups()
        state.items.append(
            build_parsed_item(item_type, item_name, state.namespace, line_hash)
        )
    if start_block_match:
        [block_type, block_name] = start_block_match.groups()
        # weird keyword `section end` in https://github.com/leanprover-community/mathlib/blob/master/src/algebraic_geometry/pullbacks.lean#L113
        # not sure what this means, so just skip it
        if block_type == "section" and block_name == "end":
            return
        state.blocks.append((cast(BlockType, block_type), block_name))
    if end_block_match:
        [block_name] = end_block_match.groups()
        if len(state.blocks) > 0 and block_name == state.blocks[-1][1]:
            state.blocks.pop()


def build_parsed_item(
    item_type: str,
    item_name: str,
    namespace: list[str],
    line_hash: str,
) -> ParsedItem:
    item_name_parts = item_name.split(".")

    if item_type not in ["lemma", "theorem", "def"]:
        raise Exception(f"Invalid item type found: {item_type}")
    safe_item_type = cast(ItemType, item_type)

    # starting with _root_ means we should throw out the namespace we're in
    if item_name_parts[0] == "_root_":
        return ParsedItem(
            type=safe_item_type,
            name=item_name_parts[-1],
            namespace=item_name_parts[1:-1],
            line_hash=line_hash,
        )
    return ParsedItem(
        type=safe_item_type,
        name=item_name_parts[-1],
        namespace=namespace + item_name_parts[0:-1],
        line_hash=line_hash,
    )
