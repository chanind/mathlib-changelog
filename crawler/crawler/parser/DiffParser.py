from __future__ import annotations
from collections import defaultdict
from dataclasses import dataclass
from hashlib import md5
import logging
from typing import Literal, Optional
from git import Repo
from git.objects import Commit

from .errors import LeanParseError
from .parse_lean import ParsedItem, parse_lean


ChangeType = Literal["add", "del", "mod"]
ItemType = Literal["theorem", "def", "inductive", "structure"]


@dataclass
class ItemChange:
    change_type: ChangeType
    item_type: ItemType
    name: str
    namespace: list[str]

    @property
    def full_name(self) -> str:
        return ".".join([*self.namespace, self.name])

    @property
    def sort_key(self) -> str:
        return f"{self.full_name}:{self.change_type}:{self.change_type}"


@dataclass
class ParsedDiff:
    old_path: str | None
    new_path: str | None
    changes: list[ItemChange]

    @property
    def file_change_str(self) -> str:
        path = self.old_path or self.new_path
        if self.old_path and self.new_path and self.old_path != self.new_path:
            return f"Renamed {self.old_path} to {self.new_path}"
        elif self.old_path and self.new_path:
            return f"Modified {path}"
        elif not self.new_path:
            return f"Deleted {path}"
        return f"Added {path}"


ParseCache = dict[str, list[ParsedItem]]


class DiffParser:
    parse_cache: ParseCache
    repo: Repo

    def __init__(self, repo: Repo, parse_cache: Optional[ParseCache] = None):
        self.repo = repo
        self.parse_cache = parse_cache or {}

    def parse_items(
        self, commit_sha: str, path: str, blob_sha: str | None
    ) -> list[ParsedItem]:
        if not path or path[-5:] != ".lean":
            return []
        git_lookup_key = f"{commit_sha}:{path}"
        if git_lookup_key in self.parse_cache:
            return self.parse_cache[git_lookup_key]
        if blob_sha and blob_sha in self.parse_cache:
            return self.parse_cache[blob_sha]
        contents = self.repo.git.show(git_lookup_key)
        contents_hash = md5(contents.encode()).hexdigest()
        if contents_hash in self.parse_cache:
            return self.parse_cache[contents_hash]
        try:
            parse_result = parse_lean(contents)
            self.parse_cache[git_lookup_key] = parse_result
            self.parse_cache[contents_hash] = parse_result
            if blob_sha:
                self.parse_cache[blob_sha] = parse_result
            return parse_result
        except LeanParseError as e:
            logging.error(f"failed parsing: {git_lookup_key}")
            raise e

    def diff_commits(self, old_commit: Commit, new_commit: Commit) -> list[ParsedDiff]:
        parsed_diffs = []
        for diff in old_commit.diff(new_commit):
            old_path = None if diff.new_file else diff.a_path
            new_path = None if diff.deleted_file else diff.b_path
            old_items = (
                self.parse_items(
                    old_commit.hexsha,
                    old_path,
                    diff.a_blob.hexsha if diff.a_blob else None,
                )
                if old_path
                else []
            )
            new_items = (
                self.parse_items(
                    new_commit.hexsha,
                    new_path,
                    diff.b_blob.hexsha if diff.b_blob else None,
                )
                if new_path
                else []
            )
            old_items_index = index_parsed_items(old_items)
            new_items_index = index_parsed_items(new_items)
            changes: list[ItemChange] = []
            for old_name, old_items in old_items_index.items():
                if old_name not in new_items_index:
                    changes.append(
                        ItemChange(
                            "del",
                            old_items[0].type,
                            old_items[0].name,
                            old_items[0].namespace,
                        )
                    )
                else:
                    new_items = new_items_index[old_name]
                    old_hashes = {item.line_hash for item in old_items}
                    new_hashes = {item.line_hash for item in new_items}
                    if old_hashes != new_hashes:
                        changes.append(
                            ItemChange(
                                "mod",
                                old_items[0].type,
                                old_items[0].name,
                                old_items[0].namespace,
                            )
                        )
            for new_name, new_items in new_items_index.items():
                if new_name not in old_items_index:
                    changes.append(
                        ItemChange(
                            "add",
                            new_items[0].type,
                            new_items[0].name,
                            new_items[0].namespace,
                        )
                    )
            sorted_changes = sorted(changes, key=lambda change: change.sort_key)
            parsed_diffs.append(
                ParsedDiff(old_path=old_path, new_path=new_path, changes=sorted_changes)
            )
        return parsed_diffs


def index_parsed_items(parsed_items: list[ParsedItem]) -> dict[str, list[ParsedItem]]:
    index = defaultdict(list)
    for item in parsed_items:
        index[item.full_name].append(item)
    return index
