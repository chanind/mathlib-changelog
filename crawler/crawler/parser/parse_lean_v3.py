from __future__ import annotations
import logging
import re
from typing import cast
import hashlib

from crawler.parser.types import BlockType, ItemType, ParseState, ParsedItem

from .strip_lean_comments import strip_lean_comments


def parse_lean_v3(contents: str) -> list[ParsedItem]:
    state = ParseState(blocks=[], items=[])

    lines = strip_lean_comments(contents).splitlines()
    for line in lines:
        parse_line(line, state)
    if len(state.namespace) > 0:
        logging.warning(
            f"Invalid parse state! open namespace remaining at end of parse: {state.namespace}"
        )
    return state.items


OPTIONAL_DECORATOR_REGEX = r"(?:@\[[^\]]+\]\s?)"
ITEM_REGEX = rf"^\s*{OPTIONAL_DECORATOR_REGEX}*(def|lemma|theorem|abbreviation|inductive|structure)\s+([^\s]+)"
START_BLOCK_REGEX = r"^\s*(section|namespace)\s+([^\s]+)"
END_BLOCK_REGEX = r"^\s*end\s+([^\s]+)"


def remap_item_type(item_type: str) -> ItemType:
    """
    map lemma -> theorem, and abbreviation -> def
    """
    if item_type == "lemma":
        return "theorem"
    if item_type == "abbreviation":
        return "def"
    return cast(ItemType, item_type)


def parse_line(line: str, state: ParseState) -> None:
    """
    Extract any items from the line, and track sections/namespaces.
    This will modify the state in-place.

    NOTE: this assumes there can only be a single def/theorem/lemma/section/namespace per line
    """
    item_match = re.search(ITEM_REGEX, line)
    start_block_match = re.search(START_BLOCK_REGEX, line)
    end_block_match = re.search(END_BLOCK_REGEX, line)

    if item_match:
        line_hash = hashlib.md5(line.encode()).hexdigest()
        [item_type, item_name] = item_match.groups()
        state.items.append(
            build_parsed_item(
                remap_item_type(item_type), item_name, state.namespace, line_hash
            )
        )
    if start_block_match:
        [block_type, block_name] = start_block_match.groups()
        # weird keyword `section end` in https://github.com/leanprover-community/mathlib/blob/master/src/algebraic_geometry/pullbacks.lean#L113
        # not sure what this means, so just skip it
        if block_type == "section" and block_name == "end":
            return
        state.blocks.append((cast(BlockType, block_type), block_name))
    if end_block_match:
        [block_name] = end_block_match.groups()
        if len(state.blocks) > 0 and block_name == state.blocks[-1][1]:
            state.blocks.pop()


def build_parsed_item(
    item_type: str,
    item_name: str,
    namespace: list[str],
    line_hash: str,
) -> ParsedItem:
    item_name_parts = item_name.split(".")
    if item_type not in [
        "theorem",
        "def",
        "inductive",
        "structure",
    ]:
        raise Exception(f"Invalid item type found: {item_type}")
    safe_item_type = cast(ItemType, item_type)

    # starting with _root_ means we should throw out the namespace we're in
    if item_name_parts[0] == "_root_":
        return ParsedItem(
            type=safe_item_type,
            name=item_name_parts[-1],
            namespace=item_name_parts[1:-1],
            line_hash=line_hash,
        )
    return ParsedItem(
        type=safe_item_type,
        name=item_name_parts[-1],
        namespace=namespace + item_name_parts[0:-1],
        line_hash=line_hash,
    )
