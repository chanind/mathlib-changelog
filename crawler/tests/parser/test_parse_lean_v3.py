from pathlib import Path
from crawler.parser.parse_lean_v3 import parse_lean_v3
from syrupy.assertion import SnapshotAssertion


FIXTURES_DIR = Path(__file__).resolve().parent / "../fixtures_v3"


def read_fixture(filename: str) -> str:
    with open(FIXTURES_DIR / filename, "r") as file:
        return file.read()


def test_parse_lean_ess_sup_fixture(snapshot: SnapshotAssertion) -> None:
    contents = read_fixture("ess_sup.lean")
    parsed_items = parse_lean_v3(contents)
    assert [item.full_name for item in parsed_items] == snapshot


def test_parse_lean_filtered_colimits_fixture(snapshot: SnapshotAssertion) -> None:
    contents = read_fixture("filtered_colimits.lean")
    parsed_items = parse_lean_v3(contents)
    assert [item.full_name for item in parsed_items] == snapshot


def test_parse_lean_algebra_fixture(snapshot: SnapshotAssertion) -> None:
    contents = read_fixture("algebra.lean")
    parsed_items = parse_lean_v3(contents)
    assert [item.full_name for item in parsed_items] == snapshot


def test_parse_lean_cont_mdiff_fixture(snapshot: SnapshotAssertion) -> None:
    contents = read_fixture("cont_mdiff.lean")
    parsed_items = parse_lean_v3(contents)
    assert [item.full_name for item in parsed_items] == snapshot


def test_parse_lean_to_additive_fixture(snapshot: SnapshotAssertion) -> None:
    contents = read_fixture("to_additive.lean")
    parsed_items = parse_lean_v3(contents)
    assert [item.full_name for item in parsed_items] == snapshot


def test_parse_lean_fermat4_fixture(snapshot: SnapshotAssertion) -> None:
    contents = read_fixture("fermat4.lean")
    parsed_items = parse_lean_v3(contents)
    assert [item.full_name for item in parsed_items] == snapshot
