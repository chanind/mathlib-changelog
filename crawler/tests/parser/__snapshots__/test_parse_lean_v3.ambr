# name: test_parse_lean_algebra_fixture
  list([
    'structure:category_theory.monad.algebra',
    'structure:category_theory.monad.algebra.hom',
    'def:category_theory.monad.algebra.hom.id',
    'def:category_theory.monad.algebra.hom.comp',
    'theorem:category_theory.monad.algebra.comp_eq_comp',
    'theorem:category_theory.monad.algebra.id_eq_id',
    'theorem:category_theory.monad.algebra.id_f',
    'theorem:category_theory.monad.algebra.comp_f',
    'def:category_theory.monad.algebra.iso_mk',
    'def:category_theory.monad.forget',
    'def:category_theory.monad.free',
    'def:category_theory.monad.adj',
    'theorem:category_theory.monad.algebra_iso_of_iso',
    'theorem:category_theory.monad.algebra_epi_of_epi',
    'theorem:category_theory.monad.algebra_mono_of_mono',
    'theorem:category_theory.monad.left_adjoint_forget',
    'theorem:category_theory.monad.of_right_adjoint_forget',
    'def:category_theory.monad.algebra_functor_of_monad_hom',
    'def:category_theory.monad.algebra_functor_of_monad_hom_id',
    'def:category_theory.monad.algebra_functor_of_monad_hom_comp',
    'def:category_theory.monad.algebra_functor_of_monad_hom_eq',
    'def:category_theory.monad.algebra_equiv_of_iso_monads',
    'theorem:category_theory.monad.algebra_equiv_of_iso_monads_comp_forget',
    'structure:category_theory.comonad.coalgebra',
    'structure:category_theory.comonad.coalgebra.hom',
    'def:category_theory.comonad.coalgebra.hom.id',
    'def:category_theory.comonad.coalgebra.hom.comp',
    'theorem:category_theory.comonad.coalgebra.comp_eq_comp',
    'theorem:category_theory.comonad.coalgebra.id_eq_id',
    'theorem:category_theory.comonad.coalgebra.id_f',
    'theorem:category_theory.comonad.coalgebra.comp_f',
    'def:category_theory.comonad.coalgebra.iso_mk',
    'def:category_theory.comonad.forget',
    'def:category_theory.comonad.cofree',
    'def:category_theory.comonad.adj',
    'theorem:category_theory.comonad.coalgebra_iso_of_iso',
    'theorem:category_theory.comonad.algebra_epi_of_epi',
    'theorem:category_theory.comonad.algebra_mono_of_mono',
    'theorem:category_theory.comonad.right_adjoint_forget',
    'theorem:category_theory.comonad.of_left_adjoint_forget',
  ])
# ---
# name: test_parse_lean_cont_mdiff_fixture
  list([
    'def:cont_diff_within_at_prop',
    'theorem:cont_diff_within_at_local_invariant_prop',
    'theorem:cont_diff_within_at_local_invariant_prop_mono',
    'theorem:cont_diff_within_at_local_invariant_prop_id',
    'def:cont_mdiff_within_at',
    'def:smooth_within_at',
    'def:cont_mdiff_at',
    'def:smooth_at',
    'def:cont_mdiff_on',
    'def:smooth_on',
    'def:cont_mdiff',
    'def:smooth',
    'theorem:cont_mdiff.smooth',
    'theorem:smooth.cont_mdiff',
    'theorem:cont_mdiff_on.smooth_on',
    'theorem:smooth_on.cont_mdiff_on',
    'theorem:cont_mdiff_at.smooth_at',
    'theorem:smooth_at.cont_mdiff_at',
    'theorem:cont_mdiff_within_at.smooth_within_at',
    'theorem:smooth_within_at.cont_mdiff_within_at',
    'theorem:cont_mdiff.cont_mdiff_at',
    'theorem:smooth.smooth_at',
    'theorem:cont_mdiff_within_at_univ',
    'theorem:smooth_at_univ',
    'theorem:cont_mdiff_on_univ',
    'theorem:smooth_on_univ',
    'theorem:cont_mdiff_within_at_iff',
    "theorem:cont_mdiff_within_at_iff''",
    'theorem:cont_mdiff_within_at_iff_target',
    'theorem:smooth_within_at_iff',
    'theorem:smooth_within_at_iff_target',
    'theorem:cont_mdiff_at_ext_chart_at',
    "theorem:cont_mdiff_within_at_iff'",
    "theorem:cont_mdiff_at_ext_chart_at'",
    'theorem:cont_mdiff_on_iff',
    'theorem:cont_mdiff_on_iff_target',
    'theorem:smooth_on_iff',
    'theorem:smooth_on_iff_target',
    'theorem:cont_mdiff_iff',
    'theorem:cont_mdiff_iff_target',
    'theorem:smooth_iff',
    'theorem:smooth_iff_target',
    'theorem:cont_mdiff_within_at.of_le',
    'theorem:cont_mdiff_at.of_le',
    'theorem:cont_mdiff_on.of_le',
    'theorem:cont_mdiff.of_le',
    'theorem:cont_mdiff_within_at.of_succ',
    'theorem:cont_mdiff_at.of_succ',
    'theorem:cont_mdiff_on.of_succ',
    'theorem:cont_mdiff.of_succ',
    'theorem:cont_mdiff_within_at.continuous_within_at',
    'theorem:cont_mdiff_at.continuous_at',
    'theorem:cont_mdiff_on.continuous_on',
    'theorem:cont_mdiff.continuous',
    'theorem:cont_mdiff_within_at.mdifferentiable_within_at',
    'theorem:cont_mdiff_at.mdifferentiable_at',
    'theorem:cont_mdiff_on.mdifferentiable_on',
    'theorem:cont_mdiff.mdifferentiable',
    'theorem:smooth.mdifferentiable',
    'theorem:smooth.mdifferentiable_at',
    'theorem:smooth.mdifferentiable_within_at',
    'theorem:cont_mdiff_within_at_top',
    'theorem:cont_mdiff_at_top',
    'theorem:cont_mdiff_on_top',
    'theorem:cont_mdiff_top',
    'theorem:cont_mdiff_within_at_iff_nat',
    'theorem:cont_mdiff_within_at.mono',
    'theorem:cont_mdiff_at.cont_mdiff_within_at',
    'theorem:smooth_at.smooth_within_at',
    'theorem:cont_mdiff_on.mono',
    'theorem:cont_mdiff.cont_mdiff_on',
    'theorem:smooth.smooth_on',
    "theorem:cont_mdiff_within_at_inter'",
    'theorem:cont_mdiff_within_at_inter',
    'theorem:cont_mdiff_within_at.cont_mdiff_at',
    'theorem:smooth_within_at.smooth_at',
    'theorem:cont_mdiff_on_ext_chart_at',
    'theorem:cont_mdiff_within_at_iff_cont_mdiff_on_nhds',
    'theorem:cont_mdiff_at_iff_cont_mdiff_on_nhds',
    'theorem:cont_mdiff_within_at.congr',
    'theorem:cont_mdiff_within_at_congr',
    'theorem:cont_mdiff_within_at.congr_of_eventually_eq',
    'theorem:filter.eventually_eq.cont_mdiff_within_at_iff',
    'theorem:cont_mdiff_at.congr_of_eventually_eq',
    'theorem:filter.eventually_eq.cont_mdiff_at_iff',
    'theorem:cont_mdiff_on.congr',
    'theorem:cont_mdiff_on_congr',
    'theorem:cont_mdiff_on_of_locally_cont_mdiff_on',
    'theorem:cont_mdiff_of_locally_cont_mdiff_on',
    'theorem:cont_mdiff_within_at.comp',
    'theorem:cont_mdiff_on.comp',
    "theorem:cont_mdiff_on.comp'",
    'theorem:cont_mdiff.comp',
    "theorem:cont_mdiff_within_at.comp'",
    'theorem:cont_mdiff_at.comp_cont_mdiff_within_at',
    'theorem:cont_mdiff_at.comp',
    'theorem:cont_mdiff.comp_cont_mdiff_on',
    'theorem:smooth.comp_smooth_on',
    'theorem:cont_mdiff_on.comp_cont_mdiff',
    'theorem:smooth_on.comp_smooth',
    'theorem:cont_mdiff_on_of_mem_maximal_atlas',
    'theorem:cont_mdiff_on_symm_of_mem_maximal_atlas',
    'theorem:cont_mdiff_on_chart',
    'theorem:cont_mdiff_on_chart_symm',
    'theorem:cont_mdiff_id',
    'theorem:smooth_id',
    'theorem:cont_mdiff_on_id',
    'theorem:smooth_on_id',
    'theorem:cont_mdiff_at_id',
    'theorem:smooth_at_id',
    'theorem:cont_mdiff_within_at_id',
    'theorem:smooth_within_at_id',
    'theorem:cont_mdiff_const',
    'theorem:cont_mdiff_one',
    'theorem:smooth_const',
    'theorem:smooth_one',
    'theorem:cont_mdiff_on_const',
    'theorem:cont_mdiff_on_one',
    'theorem:smooth_on_const',
    'theorem:smooth_on_one',
    'theorem:cont_mdiff_at_const',
    'theorem:cont_mdiff_at_one',
    'theorem:smooth_at_const',
    'theorem:smooth_at_one',
    'theorem:cont_mdiff_within_at_const',
    'theorem:cont_mdiff_within_at_one',
    'theorem:smooth_within_at_const',
    'theorem:smooth_within_at_one',
    'theorem:cont_mdiff_of_support',
    'theorem:cont_mdiff_within_at_iff_cont_diff_within_at',
    'theorem:cont_mdiff_at_iff_cont_diff_at',
    'theorem:cont_mdiff_on_iff_cont_diff_on',
    'theorem:cont_mdiff_iff_cont_diff',
    'theorem:cont_mdiff_on.continuous_on_tangent_map_within_aux',
    'theorem:cont_mdiff_on.cont_mdiff_on_tangent_map_within_aux',
    'theorem:cont_mdiff_on.cont_mdiff_on_tangent_map_within',
    'theorem:cont_mdiff_on.continuous_on_tangent_map_within',
    'theorem:cont_mdiff.cont_mdiff_tangent_map',
    'theorem:cont_mdiff.continuous_tangent_map',
    'theorem:basic_smooth_vector_bundle_core.cont_mdiff_proj',
    'theorem:basic_smooth_vector_bundle_core.smooth_proj',
    'theorem:basic_smooth_vector_bundle_core.cont_mdiff_on_proj',
    'theorem:basic_smooth_vector_bundle_core.smooth_on_proj',
    'theorem:basic_smooth_vector_bundle_core.cont_mdiff_at_proj',
    'theorem:basic_smooth_vector_bundle_core.smooth_at_proj',
    'theorem:basic_smooth_vector_bundle_core.cont_mdiff_within_at_proj',
    'theorem:basic_smooth_vector_bundle_core.smooth_within_at_proj',
    'theorem:basic_smooth_vector_bundle_core.smooth_const_section',
    'theorem:tangent_bundle.cont_mdiff_proj',
    'theorem:tangent_bundle.smooth_proj',
    'theorem:tangent_bundle.cont_mdiff_on_proj',
    'theorem:tangent_bundle.smooth_on_proj',
    'theorem:tangent_bundle.cont_mdiff_at_proj',
    'theorem:tangent_bundle.smooth_at_proj',
    'theorem:tangent_bundle.cont_mdiff_within_at_proj',
    'theorem:tangent_bundle.smooth_within_at_proj',
    'def:tangent_bundle.zero_section',
    'theorem:tangent_bundle.smooth_zero_section',
    'theorem:tangent_bundle.tangent_map_tangent_bundle_pure',
    'theorem:cont_mdiff_within_at.prod_mk',
    'theorem:cont_mdiff_within_at.prod_mk_space',
    'theorem:cont_mdiff_at.prod_mk',
    'theorem:cont_mdiff_at.prod_mk_space',
    'theorem:cont_mdiff_on.prod_mk',
    'theorem:cont_mdiff_on.prod_mk_space',
    'theorem:cont_mdiff.prod_mk',
    'theorem:cont_mdiff.prod_mk_space',
    'theorem:smooth_within_at.prod_mk',
    'theorem:smooth_within_at.prod_mk_space',
    'theorem:smooth_at.prod_mk',
    'theorem:smooth_at.prod_mk_space',
    'theorem:smooth_on.prod_mk',
    'theorem:smooth_on.prod_mk_space',
    'theorem:smooth.prod_mk',
    'theorem:smooth.prod_mk_space',
    'theorem:cont_mdiff_within_at_fst',
    'theorem:cont_mdiff_at_fst',
    'theorem:cont_mdiff_on_fst',
    'theorem:cont_mdiff_fst',
    'theorem:smooth_within_at_fst',
    'theorem:smooth_at_fst',
    'theorem:smooth_on_fst',
    'theorem:smooth_fst',
    'theorem:cont_mdiff_within_at_snd',
    'theorem:cont_mdiff_at_snd',
    'theorem:cont_mdiff_on_snd',
    'theorem:cont_mdiff_snd',
    'theorem:smooth_within_at_snd',
    'theorem:smooth_at_snd',
    'theorem:smooth_on_snd',
    'theorem:smooth_snd',
    'theorem:smooth_iff_proj_smooth',
    "theorem:cont_mdiff_within_at.prod_map'",
    'theorem:cont_mdiff_within_at.prod_map',
    'theorem:cont_mdiff_at.prod_map',
    "theorem:cont_mdiff_at.prod_map'",
    'theorem:cont_mdiff_on.prod_map',
    'theorem:cont_mdiff.prod_map',
    'theorem:smooth_within_at.prod_map',
    'theorem:smooth_at.prod_map',
    'theorem:smooth_on.prod_map',
    'theorem:smooth.prod_map',
    'theorem:cont_mdiff_within_at_pi_space',
    'theorem:cont_mdiff_on_pi_space',
    'theorem:cont_mdiff_at_pi_space',
    'theorem:cont_mdiff_pi_space',
    'theorem:smooth_within_at_pi_space',
    'theorem:smooth_on_pi_space',
    'theorem:smooth_at_pi_space',
    'theorem:smooth_pi_space',
    'theorem:continuous_linear_map.cont_mdiff',
    'theorem:smooth_smul',
    'theorem:smooth.smul',
    'theorem:smooth_on.smul',
    'theorem:smooth_at.smul',
  ])
# ---
# name: test_parse_lean_ess_sup_fixture
  list([
    'def:ess_sup',
    'def:ess_inf',
    'theorem:ess_sup_congr_ae',
    'theorem:ess_inf_congr_ae',
    'theorem:ess_sup_eq_Inf',
    'theorem:ess_sup_measure_zero',
    'theorem:ess_inf_measure_zero',
    'theorem:ess_sup_mono_ae',
    'theorem:ess_inf_mono_ae',
    'theorem:ess_sup_const',
    'theorem:ess_sup_le_of_ae_le',
    'theorem:ess_inf_const',
    'theorem:le_ess_inf_of_ae_le',
    'theorem:ess_sup_const_bot',
    'theorem:ess_inf_const_top',
    'theorem:order_iso.ess_sup_apply',
    'theorem:order_iso.ess_inf_apply',
    'theorem:ess_sup_mono_measure',
    "theorem:ess_sup_mono_measure'",
    'theorem:ess_inf_antitone_measure',
    'theorem:ess_sup_smul_measure',
    'theorem:ess_sup_comp_le_ess_sup_map_measure',
    'theorem:measurable_embedding.ess_sup_map_measure',
    'theorem:ess_sup_map_measure_of_measurable',
    'theorem:ess_sup_map_measure',
    'theorem:ae_lt_of_ess_sup_lt',
    'theorem:ae_lt_of_lt_ess_inf',
    'theorem:ess_sup_indicator_eq_ess_sup_restrict',
    'theorem:ennreal.ae_le_ess_sup',
    'theorem:ennreal.ess_sup_eq_zero_iff',
    'theorem:ennreal.ess_sup_const_mul',
    'theorem:ennreal.ess_sup_add_le',
    'theorem:ennreal.ess_sup_liminf_le',
  ])
# ---
# name: test_parse_lean_fermat4_fixture
  list([
    'def:fermat_42',
    'theorem:fermat_42.comm',
    'theorem:fermat_42.mul',
    'theorem:fermat_42.ne_zero',
    'def:fermat_42.minimal',
    'theorem:fermat_42.exists_minimal',
    'theorem:fermat_42.coprime_of_minimal',
    'theorem:fermat_42.minimal_comm',
    'theorem:fermat_42.neg_of_minimal',
    'theorem:fermat_42.exists_odd_minimal',
    'theorem:fermat_42.exists_pos_odd_minimal',
    'theorem:int.coprime_of_sq_sum',
    "theorem:int.coprime_of_sq_sum'",
    'theorem:fermat_42.not_minimal',
    'theorem:not_fermat_42',
    'theorem:not_fermat_4',
  ])
# ---
# name: test_parse_lean_filtered_colimits_fixture
  list([
    'def:SemiRing.filtered_colimits.R',
    'def:SemiRing.filtered_colimits.colimit',
    'def:SemiRing.filtered_colimits.colimit_cocone',
    'def:SemiRing.filtered_colimits.colimit_cocone_is_colimit',
    'def:CommSemiRing.filtered_colimits.R',
    'def:CommSemiRing.filtered_colimits.colimit',
    'def:CommSemiRing.filtered_colimits.colimit_cocone',
    'def:CommSemiRing.filtered_colimits.colimit_cocone_is_colimit',
    'def:Ring.filtered_colimits.R',
    'def:Ring.filtered_colimits.colimit',
    'def:Ring.filtered_colimits.colimit_cocone',
    'def:Ring.filtered_colimits.colimit_cocone_is_colimit',
    'def:CommRing.filtered_colimits.R',
    'def:CommRing.filtered_colimits.colimit',
    'def:CommRing.filtered_colimits.colimit_cocone',
    'def:CommRing.filtered_colimits.colimit_cocone_is_colimit',
  ])
# ---
# name: test_parse_lean_to_additive_fixture
  list([
    'structure:to_additive.value_type',
  ])
# ---
